lib/cryptoexchange/client.rb

```diff
@@ -80,5 +80,69 @@ def trades(market_pair)
       trades = market_class.new
       trades.fetch(market_pair)
     end
+
+    def ticker_stream(market_pair:, onopen: nil, onmessage:, onclose: nil)
+      fetch_stream(
+        market_pair: market_pair,
+        onopen: onopen,
+        onmessage: onmessage,
+        onclose: onclose,
+        stream_type: 'market'
+      )
+    end
+
+    def trade_stream(market_pair:, onopen: nil, onmessage:, onclose: nil)
+      fetch_stream(
+        market_pair: market_pair,
+        onopen: onopen,
+        onmessage: onmessage,
+        onclose: onclose,
+        stream_type: 'trade'
+      )
+    end
+
+    def order_book_stream(market_pair:, onopen: nil, onmessage:, onclose: nil)
+      fetch_stream(
+        market_pair: market_pair,
+        onopen: onopen,
+        onmessage: onmessage,
+        onclose: onclose,
+        stream_type: 'order_book'
+      )
+    end
+
+    private
+
+    def fetch_stream(market_pair:, onopen: nil, onmessage:, onclose: nil, stream_type:)
+      fail 'stream_type must be one of "market", "trade", "order_book"' unless %w(market trade order_book).include?(stream_type)
+
+      exchange         = market_pair.market
+      stream_classname = "Cryptoexchange::Exchanges::#{StringHelper.camelize(exchange)}::Services::#{StringHelper.camelize(stream_type)}Stream"
+      stream_class     = Object.const_get(stream_classname)
+      stream           = stream_class.new
+
+      Thread.new do
+        EM.run do
+          ws = WebSocket::EventMachine::Client.connect(uri: stream.url)
+
+          ws.onopen do
+            onopen.call if onopen
+            ws.send(stream.subscribe_event(market_pair))
+          end
+
+          ws.onmessage do |message, _|
+            message = JSON.parse(message)
+
```

Comment:
There should also be a way to tell why a WS socket closed in development. Otherwise developers may be lost just because their code had a typo in the block, ws closed